### PR TITLE
HW 03 by Tatsiana_Zenkovich

### PR DESCRIPTION
**Pull request by TanjaZzz**

### CODE REVIEW

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
+class Student
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:1]: There must be a separate file for each class. No need to do the whole implementation in one file, this is bad practice

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  attr_reader :name, :surname
+  def initialize(name, surname)
+    @name = name
+    @surname = surname
+    @homeworks = []
+    @mentor = []
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:7]:
```suggestion
    @homeworks = []
    @mentors = []
```

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def submit_homework!(homework_data)
+    @homeworks << homework_data
+    # puts 'homework submitted'
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:12]: You can take these conclusions to the runner file where you will display this information before calling the method or to another place

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def notify_mentor
+    @mentor.each { |mentor| mentor.update(self) }
+    # puts 'mentor notified'
+  end
+
+  def homeworks
+    puts @homeworks
+  end
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:23]: You do not need this method if you add homeworks to attr_???
Or just return homework
```suggestion
  def homeworks
    @homeworks
  end
```

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  def update(mentor)
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:25]: think about the name of the method, maybe you should call it "subscribe to the mentor"

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    @mentor << mentor
+    # puts 'mentor is added'
+  end
+end
+
+class Mentor
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:31]: Try to implement the observer pattern in both directions.
If the mentor sends homework, the student receives a notification and he begins to do homework. 2) if the student sends a decision to the mentor, then he receives a notification about it
*** 
You can enter 3 entity "task" and implement this functionality

File: HW03/Tatsiana_Zenkovich/app.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+student.homeworks # => [Homework, ...]
+
+mentor.subscribe_to(student)
+mentor.notifications # => []
+
+student.submit_homework!('homework_second')
+mentor.notifications # => [Notification, ...]
+
+mentor.read_notifications!
+mentor.notifications # => []
```
Comment (B0RlS) [inline HW03/Tatsiana_Zenkovich/app.rb:80]: Take it to a runner file

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 1 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:1]: corrected :)

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 2 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:7]: also corrected )

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 4 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:23]: corrected, as you suggested

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 7 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:80]: done! Thank you!

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 5 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:25]: I re-named it to "subscribe"

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 3 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb:12]: I used it without comment marks, when was creating the code to check, as what stage what is working or non-working. I just deleted those comments now.

File: HW03/Tatsiana_Zenkovich/app.rb (hunk 6 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/app.rb]: I made a new class Task, as you suggested, when a mentor creates and sends tasks.

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    @students << student
+  end
+
+  def unsubscribe_from(student)
+    @students.delete(student)
+  end
+
+  def read_notifications!
+    @notifications = @notifications.clear
+  end
```
Comment (vladhilko) [inline HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb:22]: Should work by this way 
```suggestion
  def read_notifications!
     @notifications.clear
  end
```

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    @surname = surname
+    @notifications = []
+    @students = []
+    @tasks = []
+  end
+
+  def subscribe_to(student)
+    @notifications.clear
+    @students << student
+  end
```
Comment (vladhilko) [inline HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb:14]: Why do you need clear `@notifications` here?

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/task.rb [hunk 10]
```diff
@@ -0,0 +1,6 @@
+class Task
+  attr_reader :title
+  def initialize(title)
```
[the remark this answers is not in the document]
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/task.rb]: done in all files!

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb (hunk 8 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb]: corrected

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb (hunk 9 above)
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb]: That was done for my first version on app.rb, when a student could send homework before mentor had subscribed to a student. So when mentor subscribed to a student, notifications were to be cleared. 
For now I just deleted it, as it's not useful anymore.

File: HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+    @students.delete(student)
+  end
+
+  def read_notifications!
+    @notifications = @notifications.clear
+  end
+
+  def send_update(_notifiaction)
+    @notifications << 'Notification'
+  end
```
[the remark this answers is not in the document]
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb]: I corrected 'Notification' to notification, 
but as for def send_update(notification) rubocop don't let the test to pass, saying "Unused method argument" and offering _notifiaction substitution. So I left it as it is.

[the remark this answers is not in the document]
Comment (vladhilko) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb]: Because you use `notifiAction` 😅

[the remark this answers is not in the document]
Comment (TanjaZzz) [reply HW03/Tatsiana_Zenkovich/tatsiana_zenkovich_hw_03_t_01/mentor.rb]: omg ))) 
corrected it! thank you!

**Pull request merged**